compiler/circle-resizer/src/ModelEditor.cpp

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2025 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ModelEditor.h"
+
+#include <mio/circle/schema_generated.h>
+
+#include <logo/Phase.h>
+#include <luci/Pass/CircleShapeInferencePass.h>
+#include <luci/Pass/CircleTypeInferencePass.h>
+#include <logo/RemoveDeadNodeWithQueryPass.h>
+#include <luci/Import/GraphBuilderRegistry.h>
+
+#include <iostream>
+#include <string>
+
+using namespace circle_resizer;
+
+namespace
+{
+void replace_tensor_shape(::flatbuffers::Vector<int32_t> *tensor_shape, const Shape &new_shape)
+{
+  const auto shape_size = tensor_shape->size();
+  if (shape_size != new_shape.rank())
+  {
+    throw std::runtime_error("Provided shape rank: " + std::to_string(new_shape.rank()) +
+                             " is different from expected: " + std::to_string(shape_size));
+  }
+  for (uint32_t dim_idx = 0; dim_idx < shape_size; ++dim_idx)
+  {
+    tensor_shape->Mutate(dim_idx, new_shape[dim_idx].value());
+  }
+}
+
+} // namespace
+
+ModelEditor::ModelEditor(std::shared_ptr<ModelData> model_data) : _model_data{model_data} {}
+
+ModelEditor &ModelEditor::resize_inputs(const std::vector<Shape> &shapes)
```

Comment:
The overall flow seems a bit complicated.

1. Import module from buffer (save them in ModelData)
2. Update buffer's input shape
3. Invalidate module
4. Import module from the updated buffer
5. Run shape inference
6. Invalidate buffer
7. Export to buffer
8. Save the buffer as a file

Do you have any reason to modify buffer's input shape? A simpler flow would be as follows.

1. Import module from buffer
2. Update module's input shape
3. Run shape inference
4. Export to buffer
5. Save the buffer as a file

Reply:
> as I know we haven't API/capabilities to update input shapes having module representation

AFAIK it is possible to update input shapes by simply updating the shapes of circle nodes in loco::Graph. The below function updates the dtype of input nodes. There are some irrelevant codes related to quantization, but you may find dtype of input node (and graph input) is updated by `dtype()` method. Similarly you can simply update shapes by `rank()` or `dim()` methods.

https://github.com/Samsung/ONE/blob/master/compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp#L384

Reply:
@jinevening Great catch, thank you! I am sure now why I haven't noticed [such](https://github.com/Samsung/ONE/pull/14727/files#diff-89115d8f21eafbf362d2bfbe1ea5677c609681b09b0ad094d6eec8a8e6e5a27eR33) option during working on the draft. To have a clear view I've closed the previous PR and create the new one (based on your simplified strategy) - https://github.com/Samsung/ONE/pull/15185

Reply:
Thanks. The code looks much better :)

---

The current tests in `one-cmds` have some issues.
1. It only checks the existence of the output file, not the updated shapes.
2. We're not yet sure about making `onecc resize`.

So, I'd suggest writing dredd tests, which we typically use for checking updates of circle models. You can find an example in `circle2circle-dredd-recipe-test`.  You may need to check the shapes of tensors in the output circle, as `test.rule` in `res/TensorFlowLiteRecipes/Inf_Mul_000`.

Reply:
@jinevening Added in https://github.com/Samsung/ONE/pull/14727/commits/f12471bd7fdd4b37c720e31901307cfc7d8e0bf4 ;-)